### PR TITLE
[linuxvm] Add a new check for LinuxVm metrics

### PR DESCRIPTION
**Pull request by DorianZaccaria**
Aggregates metrics available in /proc/vmstat and /proc

### CODE REVIEW

**Comment by irabinovitch** [comment]
Hey @DorianZaccaria  Looks like we might have some failing tests here.

**Comment by irabinovitch** [comment]
@DorianZaccaria  do we still want to keep this PR open?

File: checks.d/linuxvm.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python
```
Comment (remh) [inline checks.d/linuxvm.py]: You don't need this line

File: checks.d/linuxvm.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+"""DataDog agent checks for system.vm proc info."""
+
+from checks import AgentCheck
+
+
+def valid_instance(instance):
+    return "proc" in instance or "vmstat" in instance
```
Comment (remh) [inline checks.d/linuxvm.py]: this helper seems a bit overkill :)

File: checks.d/linuxvm.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class LinuxVm(AgentCheck):
+    VM_BASE = "system.vm"
+
+    def check(self, instance):
+        if not valid_instance(instance):
+            return
```
Comment (remh) [inline checks.d/linuxvm.py]: You should raise an Exception here

File: checks.d/linuxvm.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        proc = instance.get("proc", {})
```
Comment (remh) [inline checks.d/linuxvm.py]: you don't need to set default values as you checked before that these keys were in the instance dict.

File: checks.d/linuxvm.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        if proc:
+            for gauge in proc.get("gauges", []):
+                self.proc(self.gauge, gauge)
+
+            for counter in proc.get("counts", []):
+                self.proc(self.count, counter)
+
+        if vmstat:
+            vmstats = {x: int(y) for x, y in map(str.split, open("/proc/vmstat").readlines())}
```
Comment (remh) [inline checks.d/linuxvm.py]: python2.6 doesn't support dictionary comprehension unfortunately :(

Comment (remh) [inline checks.d/linuxvm.py]: also it's good practice to always open files in a context manager:
```python
with open('/proc/vmstat') as f:
     content = vm.readlines()
```
this way you make sure the file handler will be properly closed.

File: checks.d/linuxvm.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            print vmstats
```
Comment (remh) [inline checks.d/linuxvm.py]: should be removed.

File: checks.d/linuxvm.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            for gauge in vmstat.get("gauges", []):
+                self.vm(self.gauge, gauge, vmstats)
+
+            for counter in vmstat.get("counts", []):
+                self.vm(self.count, counter, vmstats)
+
+    def proc(self, metfunc, path):
```
Comment (remh) [inline checks.d/linuxvm.py]: you should use better names for your methods (usually use "actions")

File: checks.d/linuxvm.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        try:
+            value = float(open(path).read().strip())
```
Comment (remh) [inline checks.d/linuxvm.py]: you're not closing your file.

File: checks.d/linuxvm.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            name = path.rstrip('/').split('/')[-1]
```
Comment (remh) [inline checks.d/linuxvm.py]:
```python
import ntpath
_, name = ntpath.split(path)
```

File: checks.d/linuxvm.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            metfunc("%s.%s" % (self.VM_BASE, name), value)
+        except Exception as e:
+            self.log.error(e)
+
+    def vm(self, metfunc, stat, stats):
```
Comment (remh) [inline checks.d/linuxvm.py]: same here, it's not really understandable

File: checks.d/linuxvm.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        if stat not in stats:
+            return
```
Comment (remh) [inline checks.d/linuxvm.py]: should we log an error or something here ?

**Comment by irabinovitch** [comment]
@DorianZaccaria  any thoughts/feedbacks on Remi's comments?

**Comment by irabinovitch** [comment]
@DorianZaccaria  Ping?

**Comment by DorianZaccaria** [comment]
Not really generic, it needs a bit of work. So closing for now. We might want to revisit our approach to get these metrics.

**Comment by irabinovitch** [comment]
Thanks for the update @DorianZaccaria

**Pull request closed without merging**